lib/ruby_indexer/lib/ruby_indexer/entry.rb

```diff
@@ -69,6 +78,23 @@ class Class < Namespace
       def initialize(name, file_path, location, comments, parent_class)
         super(name, file_path, location, comments)
         @parent_class = T.let(parent_class, T.nilable(String))
+        @singleton_klass = T.let(SingletonClass.new(name, self), SingletonClass)
+      end
+    end
+
+    class SingletonClass
```

Comment:
I'm not sure about this abstraction.

In Ruby a singleton class is a class like a real instance of `Class`:

```rb
Foo.class # => Class
Foo.class.ancestors # => [Class, Module, Object, PP::ObjectMixin, Kernel, BasicObject]

Foo.class.class # => Class
Foo.class.class.ancestors # => [Class, Module, Object, PP::ObjectMixin, Kernel, BasicObject]
```

Having `SingletonClass` as a standalone will make it harder to resolve the methods properly when we call things coming from `Class` for example `Foo.name`.

I wonder if `SingletonClass` could simply inherit from `Class` or if we actually need a `SingletonClass` at all? From what I see we only have this abstraction so we store `attached_class` but it doesn't seem to be used anywhere at the moment? Or could we make `Class` old a nilable `@attached_class` and implement `singleton?` depending on if the attribute holds a value or not?

Reply:
👍 Now changed to only have `Class`.